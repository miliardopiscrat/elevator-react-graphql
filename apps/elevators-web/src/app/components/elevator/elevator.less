.elevator {
  border: 1px solid #6c757d;
  padding: 0.5em;
  border-radius: 3px;
  margin: 0.2em;
  .header {
    margin-bottom: 0.5em;
  }

  &.selected {
    border-color: #0b97c4;
    box-shadow: 0 1px 0 rgba(255, 255, 255, .7),
    inset 0 0 0 1px rgba(0, 0, 0, .1),
    inset 0 1px 1px 1px rgba(0, 0, 0, .12),
    inset 0 0 5px rgba(0, 0, 0, .1);
  }
}
