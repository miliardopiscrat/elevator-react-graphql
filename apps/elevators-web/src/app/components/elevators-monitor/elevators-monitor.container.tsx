import React, { FC } from 'react';
import { useGetElevatorsQuery } from '@elevators/elevators-api';
import { ElevatorsMonitorView } from './elevators-monitor.view';

export const ElevatorsMonitorContainer: FC = () => {
  const { loading, error, data } = useGetElevatorsQuery();

  if (error) return <p>{error.message}</p>;
  const elevators = data ? data.elevators : [];

  return <ElevatorsMonitorView isLoading={loading} elevators={elevators}/>;
};
