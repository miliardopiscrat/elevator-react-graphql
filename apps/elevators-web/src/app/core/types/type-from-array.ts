export type TypeFromArray<T> = T extends (infer U)[] ? U : never;
