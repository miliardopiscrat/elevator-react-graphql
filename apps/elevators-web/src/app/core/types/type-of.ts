export type TypeOf<T, K extends keyof T> = T[K];
