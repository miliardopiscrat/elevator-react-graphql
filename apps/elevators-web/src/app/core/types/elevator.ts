import { TypeOf } from './type-of';
import { GetElevatorsQuery } from '@elevators/elevators-api';
import { TypeFromArray } from './type-from-array';

export type ElevatorMonitorElement = TypeFromArray<TypeOf<GetElevatorsQuery, 'elevators'>>;
