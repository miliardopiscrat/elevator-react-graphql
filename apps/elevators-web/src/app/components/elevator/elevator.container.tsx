import React, { useEffect, useState, useRef, useCallback } from 'react';
import { ElevatorMonitorElement } from '../../core/types/elevator';
import { useMonitorElevatorSubscription } from '@elevators/elevators-api';
import { ElevatorView } from './elevator.view';
import { showFloorChangeSuccessNotification } from '../../core/notifications/show-floor-change-success.notification';
import { useMutation, useQuery } from '@apollo/react-hooks';
import { changeSelectedElevatorGQL } from '../../core/graph-tags/change-selected-elevator.client';
import { getSelectedElevatorGQL, SelectedElevatorIdType } from '../../core/graph-tags/get-selected-elevator.client';

type PropsType = {
  elevator: ElevatorMonitorElement;
};

export const ElevatorContainer = ({ elevator }: PropsType) => {
  const { data, error } =
    useMonitorElevatorSubscription({ variables: { id: elevator.id } });
  const [elevatorStatus, setStatus] = useState(elevator);
  const previousStatus = useRef<ElevatorMonitorElement>();
  const [changeSelection] = useMutation(changeSelectedElevatorGQL, { variables: { id: elevator.id } });
  const changeSelectionCallBack = useCallback(() => changeSelection(), [changeSelection]);
  const querySelection = useQuery<SelectedElevatorIdType, {}>(getSelectedElevatorGQL);

  useEffect(() => {
    previousStatus.current = elevatorStatus;
  }, [elevatorStatus]);
  useEffect(() => {
    if (data) setStatus(data.elevator); else setStatus(elevator);

    if (previousStatus.current && previousStatus.current.isBusy && !data.elevator.isBusy) {
      showFloorChangeSuccessNotification(elevator.id);
    }
  }, [data, elevator]);

  if (error) return <div>{error.message}</div>;

  return <ElevatorView elevator={elevatorStatus}
                       isSelected={querySelection.data && querySelection.data.selectedElevatorId === elevatorStatus.id}
                       onClick={changeSelectionCallBack}/>;
};
