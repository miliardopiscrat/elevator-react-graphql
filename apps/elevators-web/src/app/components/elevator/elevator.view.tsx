import React, { FC } from 'react';
import { ElevatorMonitorElement } from '../../core/types/elevator';
import Icon from 'antd/lib/icon';
import Statistic from 'antd/lib/statistic';

import { Direction } from '@elevators/elevators-api';
import './elevator.less';
import classNames from 'classnames';

interface PropsType {
  elevator: ElevatorMonitorElement;
  isSelected: boolean;
  onClick: () => void
}

const statisticStyle: Record<Direction, { type: string; color: string; }> = {
  [Direction.Down]: { type: 'arrow-down', color: '#cf1322' },
  [Direction.Up]: { type: 'arrow-up', color: '#cf1322' },
  [Direction.None]: { type: 'stop', color: '#72c11f' }
};

export const ElevatorView: FC<PropsType> = ({ elevator, isSelected, onClick }: PropsType) => {
  const { color, type } = statisticStyle[elevator.direction] || statisticStyle.NONE;

  return <div data-testid="elevator" className={classNames('elevator', { 'selected': isSelected })}
              onClick={onClick}>
    <div className='header'>
      <pre>Elevator no. {elevator.id}</pre>
    </div>
    <Statistic
      title="Floor"
      value={elevator.currentFloor}
      valueStyle={{ color }}
      prefix={<Icon type={type}/>}
      style={{ minWidth: '90px' }}
    />
  </div>;
};
