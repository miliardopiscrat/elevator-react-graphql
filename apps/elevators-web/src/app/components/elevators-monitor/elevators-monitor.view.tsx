import React, { FC } from 'react';
import Card from 'antd/lib/card';
import Col from 'antd/lib/col';
import Row from 'antd/lib/row';
import Skeleton from 'antd/lib/skeleton';
import Spin from 'antd/lib/spin';
import { ElevatorsView } from '../elevators/elevators.view';
import { ElevatorMonitorElement } from '../../core/types/elevator';

interface PropsType {
  isLoading: boolean;
  elevators: ElevatorMonitorElement[];
}

const pleaseWaitMessage = <Row style={{ height: '100%' }}><Col span={24}><Spin tip="Loading..."><Skeleton
  active/></Spin>
</Col></Row>;

export const ElevatorsMonitorView: FC<PropsType> = ({ isLoading, elevators }) => {

  if (isLoading) return pleaseWaitMessage;

  return  <Card title="Elevators" style={{textAlign: 'left'}}>
    <ElevatorsView elevators={elevators}/>
  </Card>;
};
