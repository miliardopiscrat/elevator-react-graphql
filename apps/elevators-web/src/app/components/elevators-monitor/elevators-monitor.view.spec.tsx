import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/react-testing';
import { ElevatorsMonitorView } from './elevators-monitor.view';
import { ElevatorMonitorElement } from '../../core/types/elevator';
import { Direction, MonitorElevatorDocument } from '@elevators/elevators-api';

describe('ElevatorsMonitorView', () => {
  afterEach(cleanup);

  it('should render successfully', () => {
    const { baseElement } = render(
      <MockedProvider addTypename={false} mocks={[]} resolvers={{}}>
        <ElevatorsMonitorView elevators={[]} isLoading={false}/>
      </MockedProvider>);

    expect(baseElement).toBeTruthy();
  });

  it('should render loading page', () => {
    const { getByText } = render(
      <MockedProvider addTypename={false} mocks={[]} resolvers={{}}>
        <ElevatorsMonitorView elevators={[]} isLoading={true}/>
      </MockedProvider>);

    expect(getByText('Loading...')).toBeInTheDocument();
  });

  it('should render elevator', () => {
    const subscription: MockedResponse = {
      request: {
        query: MonitorElevatorDocument,
        variables: {id: 12}
      },
      result: {
        data: {
          elevator: {
            id: 12,
            addressedFloor: 12,
            isBusy: false,
            direction: Direction.Up,
            currentFloor: 12
          } as ElevatorMonitorElement
        }
      }
    };

    const { getByTestId } = render(
      <MockedProvider addTypename={false} mocks={[subscription]} resolvers={{}}>
        <ElevatorsMonitorView elevators={[{id: 12} as ElevatorMonitorElement]} isLoading={false}/>
      </MockedProvider>);

    expect(getByTestId('elevator')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(<MockedProvider addTypename={false} mocks={[]} resolvers={{}}>
        <ElevatorsMonitorView elevators={[]} isLoading={false}/>
      </MockedProvider>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
