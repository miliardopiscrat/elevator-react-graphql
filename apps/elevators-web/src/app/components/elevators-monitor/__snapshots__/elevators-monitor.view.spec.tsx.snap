// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElevatorsMonitorView should match snapshot 1`] = `"<div style=\\"text-align: left;\\" class=\\"ant-card ant-card-bordered\\"><div class=\\"ant-card-head\\"><div class=\\"ant-card-head-wrapper\\"><div class=\\"ant-card-head-title\\">Elevators</div></div></div><div class=\\"ant-card-body\\"><div class=\\"elevators-container\\"></div></div></div>"`;
